### ISSUE TITLE
Automatically Aligned Labels outside Figure with Constrained Layout in Exported File

### ISSUE THREAD

**Issue by Firionus** [matplotlib/matplotlib#19152]
### Bug report

**Bug summary**

When using constrained layout on a specific 3-plot layout with automatically aligned labels, the exported plot has labels outside the page. The image in Jupyter Lab renders fine. 

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

data = [0,1]

fig, (ax3, ax1, ax2) = plt.subplots(3, 1, constrained_layout=True, figsize=(6.4,8), 
    gridspec_kw={"height_ratios":(1, 1, .7)})

ax1.plot(data, data)
ax1.set_ylabel("Label")

ax2.plot(data, data)
ax2.set_ylim(-1.5, 1.5)
ax2.set_ylabel("Label")

ax3.plot(data, data)
ax3.set_ylabel("Label")

fig.align_ylabels(axs=(ax3, ax1, ax2))

fig.savefig("test.pdf")
```

In `savefig`, `backend="PDF"` and `backend="PGF"` or exporting as .png leads to the same, unexpected outcome for me. 

**Actual outcome**

Exported PDF Screenshot: 

![image](https://user-images.githubusercontent.com/32422182/102675910-7eeb0580-419b-11eb-9bbd-2e2987c9b793.png)

**Expected outcome**

Screenshot from Jupyter Notebook: 

![image](https://user-images.githubusercontent.com/32422182/102675927-92966c00-419b-11eb-964b-cf201a1f06fd.png)

**Matplotlib version**

  * Operating system: Windows 10 2004 Build 19041.685
  * Matplotlib version: 3.3.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.7.3
  * Jupyter version (if applicable):
    * jupyter core     : 4.6.3
    * jupyter-notebook : 6.1.1
    * qtconsole        : 4.7.5
    * ipython          : 7.17.0
    * ipykernel        : 5.3.4
    * jupyter client   : 6.1.6
    * jupyter lab      : 2.2.4
    * nbconvert        : 5.6.1
    * ipywidgets       : 7.5.1
    * nbformat         : 5.0.7
    * traitlets        : 4.3.3
  * Other libraries: 

Matplotlib installed via pip. Python installed via Windows installer from python.org. 

**Workaround**

Use manual label positions as explained in https://matplotlib.org/gallery/pyplots/align_ylabels.html

**Comment by jklymak** [comment]
Thanks for the bug report.  This (coincidentally) appears to have been fixed by #17494 and will be in the 3.4 release, so I don't think there is an action item here, except perhaps to add a test for this?  I'll add a "Good first issue" here in case someone wants to add a test.  Probably can be done w/o an image comparison, but I'd have to play with it to be sure.

**Comment by Username8192** [comment]
changing `ax2.set_ylim(-1.5, 1.5)`

to `ax2.set_ylim(1.5)` solves this issue?

**Comment by Firionus** [comment]
Yes, changing to `ax2.set_ylim(1.5)` solves the issue, but also changes the axis limits, which may not be desired. The layout problem is rather sensitive to the content of the y-axis labels.

**Comment by Hathaway2010** [comment]
I'm brand new to open source — if I were looking to write a test, what files should I look at? Thank you!

**Comment by jklymak** [comment]
@Hathway2010 Please have a look at https://matplotlib.org/devel/index.html and https://github.com/matplotlib/matplotlib/tree/master/lib/matplotlib/tests  I think maybe `test_figure.py` is where the `align_ylabel` tests are?  Thanks!

**Comment by Hathaway2010** [comment]
Thank you! I'll see if I can do something in the next few days!

**Comment by Hathaway2010** [comment]
Hi! I'm pretty sure I've found a test that does the trick: after saving the pdf in the code for reproduction, I do

`after_align = (ax1.yaxis.label.get_window_extent(),
                       ax2.yaxis.label.get_window_extent(),
                       ax3.yaxis.label.get_window_extent())
  assert round(after_align[0].x0, 5) == round(after_align[1].x0, 5) == round(after_align[2].x0, 5) > 1`

When I use an old version of matplotlib, this detects that the top two labels are 1) not aligned with the third (i.e., the three aren't approximately equal) and 2) have gone off the page (have an x0 greater than 1). When running pytest on my branch of the current repository, everything seems to be fine — the test passes, and the saved PDF looks fine.

I have two beginner's questions:

1) On https://matplotlib.org/3.2.1/devel/testing.html it says "Some tests have internal side effects that need to be cleaned up after their execution (such as created figures or modified rc params). The pytest fixture mpl_test_settings() will automatically clean these up; there is no need to do anything further." When I run this test using pytest, at least singly instead of as part of the test suite, I _am_ left with that PDF. Is there a particular path to which the PDF should be saved so it will be cleaned up afterward, or something else I need to do to make sure it doesn't hang around?

2. You pointed me (with some uncertainty) to test_figure.py. However, I see there's also a test_constrainedlayout.py — would it be more appropriate to put my test in the latter document? (I do see a few tests in test_figure.py that include constrained layout)

Many thanks, and sorry for my ignorance.

**Comment by Firionus** [comment]
Hey,
thanks for taking this on!

I think your code example has lost some linebreaks. If you want line breaks in code, use a [Fenced Code Block](https://github.github.com/gfm/#fenced-code-blocks) with three backticks or tildes. 

Example:

    ``` 
    line1
    line2
    ``` 

renders as
```
line1
line2
```

**Comment by Hathaway2010** [comment]
fixed, thank you!

**Comment by jklymak** [comment]
Does this need to be saved as pdf?  Or do you just need to trigger a draw via `fig.canvas.draw()`

Yes, it seems `test_constrained_layout.py` is a good place to put this.

**Comment by Hathaway2010** [comment]
`fig.canvas.draw()` seems to do just fine, thank you and oops! I've moved the test to the correct file and hope to submit a pull request within the hour.

**Comment by jklymak** [comment]
Closed by #19373
